"""
A single cell ATAC-seq analysis pipeline.
"""

configfile: "config.yaml"

import yaml
import sys
import os

from MAESTRO.scATAC_utility import get_fastqlist, ENV_PATH, SCRIPT_PATH, RSCRIPT_PATH

def qcplot_input(wildcards):
    checkpoint_output = checkpoints.scatac_samsplit.get(**wildcards).output[0]
    return expand("Result/Log/bamLog/{barcodeid}.mapping.log",
           barcodeid=glob_wildcards(os.path.join(checkpoint_output, "{barcodeid}.sam")).barcodeid)

rule all:
    input:
        summaryreport = "Result/" + config["outprefix"] + "_scATAC_report.html"
        # tfscore = "Result/Analysis/" + config["outprefix"] + "_tfscore.txt"

if config["platform"] == "microfluidic":
    rule scatac_map:
        input:
            fasta = config["genome"]["fasta"],
            fastq1 = config["fastqdir"] + "/{fastqid}_1.fastq",
            fastq2 = config["fastqdir"] + "/{fastqid}_2.fastq"
        output:
            bam = "Result/BWA/{fastqid}.sortedByPos.bam"
        threads:
            config["cores"]
        shell:
            "bwa mem -t {threads} {input.fasta} {input.fastq1} {input.fastq2} "
            "| samtools view --threads {threads} -b "
            "| samtools sort --threads {threads} -o {output.bam}"

    rule scatac_bamrmdp:
        input:
            bam = "Result/BWA/{fastqid}.sortedByPos.bam"
        output:
            bam = "Result/BWA/{fastqid}.sortedByPos.rmdp.bam",
            metric = "Result/BWA/{fastqid}.sortedByPos.rmdp.txt"
        shell:
            "picard MarkDuplicates INPUT={input.bam} OUTPUT={output.bam} METRICS_FILE={output.metric};"
            "rm {input.bam}"

    # rule scatac_bamindex:
    #     input:
    #         bam = "Result/BWA/{fastqid}.sortedByPos.rmdp.bam"
    #     output:
    #         bai = "Result/BWA/{fastqid}.sortedByPos.rmdp.bam.bai"
    #     shell:
    #         "samtools index {input.bam}"

    rule scatac_bammerge:
        input:
            bam = expand("Result/BWA/{fastqid}.sortedByPos.rmdp.bam", fastqid = get_fastqlist(config["fastqdir"]))
        output:
            bam = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.bam",
            bamlist = "Result/BWA/" + config["outprefix"] + "_bamlist.txt",
            fragbed = "Result/QC/" + config["outprefix"] + "_frag.bed"
        params:
            sam = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.sample.sam",
            bamprefix = "Result/BWA/" + config["outprefix"] + "_bamlist_",
            subprefix = "Result/BWA/" + config["outprefix"]
        threads:
            config["cores"]
        shell:
            "ls Result/BWA/*.sortedByPos.rmdp.bam > {output.bamlist};"
            "split -1000 -d {output.bamlist} {params.bamprefix};"
            "for file in $(ls {params.bamprefix}*); do sub=${{file#{params.bamprefix}}};"
            "samtools merge --threads {threads} {params.subprefix}.${{sub}}.sortedByPos.rmdp.bam -b ${{file}}; done;"
            "samtools merge --threads {threads} {output.bam} {params.subprefix}.*.sortedByPos.rmdp.bam;"
            "rm {params.subprefix}.[0-9]*.sortedByPos.rmdp.bam;"
            "samtools view -@ {threads} -s 0.01 -o {params.sam} {output.bam};"
            "awk '{{if ($9>0) print $9}}' {params.sam} > {output.fragbed};"

    rule scatac_single_qcstat:
        input:
            bam = "Result/BWA/{fastqid}.sortedByPos.rmdp.bam",
            promoter = SCRIPT_PATH + "/annotations/" + config["species"] + "_promoter.bed",
            chrM = SCRIPT_PATH + "/annotations/" + config["species"] + "_chrM.bed",
            peak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak"
        output:
            log = "Result/Log/bamLog/{fastqid}.mapping.log",
            bam = "Result/BWA/{fastqid}.sortedByPos.rmdp.unique.bam",
            bed = "Result/BWA/{fastqid}.sortedByPos.rmdp.unique.bed",
        shell:
            "samtools flagstat {input.bam} > {output.log};"
            "samtools view -h {input.bam} | grep -v -e 'XA:Z:' -e 'SA:Z:' | samtools view -F 0x900 -f 0x2 -q 5 -b -o {output.bam};"
            "samtools view {output.bam} -c >> {output.log};"
            "bamToBed -i {output.bam} > {output.bed};"
            "grep 'chrM' {output.bed} -c >> {output.log} || true;"
            "cut -f 1-3 {output.bed} | grep -v 'chrM' | intersectBed -wa -a - -b {input.promoter} -u | wc -l >> {output.log} || true;"
            "cut -f 1-3 {output.bed} | grep -v 'chrM' | intersectBed -wa -a - -b {input.peak} -u | wc -l >> {output.log} || true ;"

    # rule scatac_bulk_unique:
    #     input:
    #         bam = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.bam"
    #     output:
    #         unique = "Result/QC/" + config["outprefix"] + "_uniquereads.txt"
    #     threads:
    #         config["cores"]
    #     shell:
    #         "samtools view -@ {threads} -h {input.bam} | grep -v -e 'XA:Z:' -e 'SA:Z:' | samtools view -@ {threads} -F 0x900 -f 0x2 -q 5 -c > {output.unique}"

    # rule scatac_bulk_qcstat:
    #     input:
    #         bam = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.bam",
    #         promoter = SCRIPT_PATH + "/annotations/" + config["species"] + "_promoter.bed",
    #         chrM = SCRIPT_PATH + "/annotations/" + config["species"] + "_chrM.bed",
    #         peak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak"
    #     output:
    #         bulkqc = "Result/QC/" + config["outprefix"] + "_bam_stat.txt",
    #         bed = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.bed",
    #     shell:
    #         "samtools flagstat {input.bam} > {output.bulkqc};"
    #         "bamToBed -i {input.bam} > {output.bed};"
    #         "intersectBed -wa -a {output.bed} -b {input.chrM} -u | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | intersectBed -wa -a - -b {input.promoter} -u | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | intersectBed -wa -a - -b {input.peak} -u | wc -l >> {output.bulkqc};"

    rule scatac_qcplot:
        input:
            log = expand("Result/Log/bamLog/{fastqid}.mapping.log", fastqid = get_fastqlist(config["fastqdir"])),
            fragbed = "Result/QC/" + config["outprefix"] + "_frag.bed",
            # unique = "Result/QC/" + config["outprefix"] + "_uniquereads.txt"
        output:
            stat = "Result/QC/singlecell.txt",
            readdistr = "Result/QC/" + config["outprefix"] + "_scATAC_read_distr.png",
            qcfrag = "Result/QC/" + config["outprefix"] + "_scATAC_fragment_size.png",
            qcmap = "Result/QC/" + config["outprefix"] + "_scATAC_mapping_summary.png",
            qcfrip = "Result/QC/" + config["outprefix"] + "_scATAC_cell_filtering.png",
            validbarcode = "Result/QC/" + config["outprefix"] + "_scATAC_validcells.txt",
        params:
            log = "Result/Log/bamLog/",
            platform = config["platform"],
            outdir = "Result/QC",
            outpre = config["outprefix"],
            stat = "singlecell.txt",
            fragbed = config["outprefix"] + "_frag.bed",
            # unique = config["outprefix"] + "_uniquereads.txt",
        shell:
            "python " + SCRIPT_PATH + "/scATAC_microfluidic_qc.py {params.log} {output.stat};"
            "Rscript " + RSCRIPT_PATH + "/scATACseq_qc.R {params.stat} {params.fragbed} {params.platform} {params.outpre} {params.outdir}"

    rule scatac_allpeakcall:
        input:
            bam = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.bam"
        output:
            peak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
            bdg = "Result/Analysis/" + config["outprefix"] + "_all_treat_pileup.bdg",
        params:
            name = config["outprefix"] + "_all"
        log:
            "Result/Log/" + config["outprefix"] + "_macs2_allpeak.log"
        shell:
            "macs2 callpeak -g hs --outdir Result/Analysis/ -n {params.name} -B -q 0.05 --nomodel --extsize=50 --SPMR -t {input.bam}"
    
    if config["shortpeaks"]:
        rule scatac_shortfragment:
            input:
                bam = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.bam"
            output:
                shortbam = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.150bp.bam"
            threads:
                config["cores"]
            shell:
                "samtools view -@ {threads} -h {input.bam} | "
                "awk -F'\\t' 'function abs(x){{return ((x < 0.0) ? -x : x)}} {{if (abs($9)<=150) print}}' | "
                "samtools view -@ {threads} -b -o {output.shortbam}"
        
        rule scatac_shortpeakcall:
            input:
                shortbam = "Result/BWA/" + config["outprefix"] + ".merged.sortedByPos.rmdp.150bp.bam"            
            output:
                bed = "Result/Analysis/" + config["outprefix"] + "_150bp_peaks.narrowPeak"
            params:
                name = config["outprefix"] + "_150bp"
            log:
                "Result/Log/" + config["outprefix"] + "_macs2_shortpeak.log"
            shell:
                "macs2 callpeak -g hs --outdir Result/Analysis -n {params.name} -B -q 0.05 --nomodel --extsize=50 --SPMR -t {input.shortbam}"
    
    if config["custompeaks"] and config["shortpeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                shortpeak = "Result/Analysis/" + config["outprefix"] + "_150bp_peaks.narrowPeak",
                custompeak = config["custompeaksloc"]
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} {input.shortpeak} {input.custompeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    elif config["custompeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                custompeaks = config["custompeaksloc"]
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"                
            shell:
                "cat {input.allpeak} {input.custompeaks} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    elif config["shortpeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                shortpeak = "Result/Analysis/" + config["outprefix"] + "_150bp_peaks.narrowPeak"
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} {input.shortpeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    else:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"

    rule scatac_countpeak:
        input:
            finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed",
            validbarcode = "Result/QC/" + config["outprefix"] + "_scATAC_validcells.txt"
        output:
            count = "Result/Analysis/" + config["outprefix"] + "_peak_count.h5"
        threads:
            config["cores"]
        params:
            bamdir = "Result/BWA",
            species = config["species"]
        shell:
            "python " + SCRIPT_PATH + "/scATAC_microfluidic_count.py {input.finalpeak} {input.validbarcode} {params.bamdir} {output.count} {threads} {params.species}"
            
if config["platform"] == "10x-genomics":
    rule scatac_cellranger:
        input:
            fastqs = config["fastqdir"],
            genome = config["genome"]["cellranger"]
        output:
            stat = "Result/Cellranger/" + config["outprefix"] + "/outs/singlecell.csv",
            bam = "Result/Cellranger/" + config["outprefix"] + "/outs/possorted_bam.bam",
            fraggz = "Result/Cellranger/" + config["outprefix"] + "/outs/fragments.tsv.gz",
            frag = "Result/Cellranger/" + config["outprefix"] + "/outs/fragments.tsv"
        params:
            fastqprefix = config["fastqprefix"],
            outpre = config["outprefix"],
            sourceoutdir = config["outprefix"] + "/",
            destoutdir = "Result/Cellranger/"
        threads:
            config["cores"]
        log:
            "Result/Log/" + config["outprefix"] + "_cellranger.log"
        shell:
            "cellranger-atac count --id={params.outpre} --fastqs={input.fastqs} "
            "--reference={input.genome} --sample={params.fastqprefix} --localcores={threads} "
            ">> {log};"
            "cp -fr {params.sourceoutdir} {params.destoutdir};"
            "rm -r {params.sourceoutdir};"
            "gzip -d {output.fraggz};"

    # rule scatac_bulk_unique:
    #     input:
    #         bam = "Result/Cellranger/" + config["outprefix"] + "/outs/possorted_bam.bam"
    #     output:
    #         unique = "Result/QC/" + config["outprefix"] + "_uniquereads.txt"
    #     threads:
    #         config["cores"]
    #     shell:
    #         "samtools view -@ {threads} -h {input.bam} | grep -v -e 'XA:Z:' -e 'SA:Z:' | samtools view -@ {threads} -F 0x900 -f 0x2 -q 5 -c > {output.unique}"

    # rule scatac_bulk_qcstat:
    #     input:
    #         bam = "Result/Cellranger/" + config["outprefix"] + "/outs/possorted_bam.bam",
    #         promoter = SCRIPT_PATH + "/annotations/" + config["species"] + "_promoter.bed",
    #         chrM = SCRIPT_PATH + "/annotations/" + config["species"] + "_chrM.bed",
    #         peak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak"
    #     output:
    #         bulkqc = "Result/QC/" + config["outprefix"] + "_bam_stat.txt",
    #         bed = "Result/Cellranger/" + config["outprefix"] + "/outs/possorted_bam.bed",
    #     shell:
    #         "samtools flagstat {input.bam} > {output.bulkqc};"
    #         "bamToBed -i {input.bam} > {output.bed};"
    #         "intersectBed -wa -a {input.bam} -b {input.chrM} -u | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | intersectBed -wa -a - -b {input.promoter} -u | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | intersectBed -wa -a - -b {input.peak} -u | wc -l >> {output.bulkqc};"

    rule scatac_qcplot:
        input:
            bam = "Result/Cellranger/" + config["outprefix"] + "/outs/possorted_bam.bam",
            stat = "Result/Cellranger/" + config["outprefix"] + "/outs/singlecell.csv",
            # unique = "Result/QC/" + config["outprefix"] + "_uniquereads.txt",
        output:
            readdistr = "Result/QC/" + config["outprefix"] + "_scATAC_read_distr.png",
            fragbed = "Result/QC/" + config["outprefix"] + "_frag.bed",
            qcfrag = "Result/QC/" + config["outprefix"] + "_scATAC_fragment_size.png",
            qcmap = "Result/QC/" + config["outprefix"] + "_scATAC_mapping_summary.png",
            qcfrip = "Result/QC/" + config["outprefix"] + "_scATAC_cell_filtering.png",
            validbarcode = "Result/QC/" + config["outprefix"] + "_scATAC_validcells.txt",
        params:
            platform = config["platform"],
            outdir = "Result/QC",
            outpre = config["outprefix"],
            stat = "../Cellranger/" + config["outprefix"] + "/outs/singlecell.csv",
            fragbed = config["outprefix"] + "_frag.bed",
            sam = "Result/Cellranger/" + config["outprefix"] + "/outs/possorted_bam_sample.sam",
            # unique = config["outprefix"] + "_uniquereads.txt",
        threads:
            config["cores"]
        shell:
            "samtools view -@ {threads} -s 0.01 -o {params.sam} {input.bam};"
            "awk '{{if ($9>0) print $9}}' {params.sam} > {output.fragbed};"
            "Rscript " + RSCRIPT_PATH + "/scATACseq_qc.R {params.stat} {params.fragbed} {params.platform} {params.outpre} {params.outdir}"

    rule scatac_allpeakcall:
        input:
            bam = "Result/Cellranger/" + config["outprefix"] + "/outs/possorted_bam.bam"
        output:
            peak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
            bdg = "Result/Analysis/" + config["outprefix"] + "_all_treat_pileup.bdg",
        params:
            name = config["outprefix"] + "_all"
        log:
            "Result/Log/" + config["outprefix"] + "_macs2_allpeak.log"
        shell:
            "macs2 callpeak -g hs --outdir Result/Analysis -n {params.name} -B -q 0.05 --nomodel --extsize=50 --SPMR -t {input.bam}"

    if config["shortpeaks"]:
        rule scatac_shortfragment:
            input:
                bam = "Result/Cellranger/" + config["outprefix"] + "/outs/possorted_bam.bam"
            output:
                shortbam = "Result/Cellranger/" + config["outprefix"] + "possorted_150bp.bam"
            threads:
                config["cores"]
            shell:
                "samtools view -@ {threads} -h {input.bam} | "
                "awk -F'\\t' 'function abs(x){{return ((x < 0.0) ? -x : x)}} {{if (abs($9)<=150) print}}' | "
                "samtools view -@ {threads} -b -o {input.shortbam}"
        
        rule scatac_shortpeakcall:
            input:
                shortbam = "Result/Cellranger/" + config["outprefix"] + "possorted_150bp.bam"
            output:
                bed = "Result/Analysis/" + config["outprefix"] + "_150bp_peaks.narrowPeak"
            params:
                name = config["outprefix"] + "_150bp"
            log:
                "Result/Log/" + config["outprefix"] + "_macs2_shortpeak.log"
            shell:
                "macs2 callpeak -g hs --outdir Result/Analysis -n {params.name} -B -q 0.05 --nomodel --extsize=50 --SPMR -t {input.shortbam}"
    
    if config["custompeaks"] and config["shortpeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                shortpeak = "Result/Analysis/" + config["outprefix"] + "_150bp_peaks.narrowPeak",
                custompeak = config["custompeaksloc"]
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} {input.shortpeak} {input.custompeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    elif config["custompeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                custompeaks = config["custompeaksloc"]
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"                
            shell:
                "cat {input.allpeak} {input.custompeaks} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    elif config["shortpeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                shortpeak = "Result/Analysis/" + config["outprefix"]  + "_150bp_peaks.narrowPeak"
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} {input.shortpeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    else:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
     
    rule scatac_countpeak:
        input:
            finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed",
            validbarcode = "Result/QC/" + config["outprefix"] + "_scATAC_validcells.txt",
            frag = "Result/Cellranger/" + config["outprefix"] + "/outs/fragments.tsv"
        output:
            count = "Result/Analysis/" + config["outprefix"] + "_peak_count.h5"
        params:
            species = config["species"]
        threads:
            config["cores"]
        shell:
            "python " + SCRIPT_PATH + "/scATAC_cellranger_count.py {input.finalpeak} {input.validbarcode} {input.frag} {output.count} {threads} {params.species}"

if config["platform"] == "sci-ATAC-seq":
    rule scatac_map:
        input:
            fasta = config["genome"]["fasta"],
            fastq1 = config["fastqdir"] + "/" + config["fastqprefix"] + "_1.fastq",
            fastq2 = config["fastqdir"] + "/" + config["fastqprefix"] + "_2.fastq"
        output:
            bam = temp("Result/BWA/" + config["outprefix"] + ".sortedByPos.bam")
        threads:
            config["cores"]
        log:
            "Result/Log/" + config["outprefix"] + "_BWA_map.log"
        shell:
            "bwa mem -t {threads} {input.fasta} {input.fastq1} {input.fastq2} "
            "| samtools view --threads {threads} -b "
            "| samtools sort --threads {threads} -o {output.bam}"

    rule scatac_rmdp:
        input:
            bam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.bam"
        output:
            bam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.bam",
            metric = "Result/BWA/" + config["outprefix"] + ".rmdp.txt",
            fragbed = "Result/QC/" + config["outprefix"] + "_frag.bed"
        params:
            sam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.sample.ÃŸsam"
        threads:
            config["cores"]
        shell:
            "picard MarkDuplicates INPUT={input.bam} OUTPUT={output.bam} METRICS_FILE={output.metric};"
            "samtools view -@ {threads} -s 0.01 -o {params.sam} {input.bam};"
            "awk '{{if ($9>0) print $9}}' {params.sam} > {output.fragbed};"

    checkpoint scatac_samsplit:
        input:
            bam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.bam",
            barcodelib = config["barcodelib"]
        output:
            bam = directory("Result/BWA/Split/"),
        threads:
            config["cores"]
        shell:
            "python " + SCRIPT_PATH + "/scATAC_sci_split.py -S {input.bam} -P {threads} -B {input.barcodelib}"

    rule scatac_single_qcstat:
        input:
            sam = "Result/BWA/Split/{barcodeid}.sam",
            promoter = SCRIPT_PATH + "/annotations/" + config["species"] + "_promoter.bed",
            # chrM = SCRIPT_PATH + "/annotations/" + config["species"] + "_chrM.bed",
            peak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak"
        output:
            log = "Result/Log/bamLog/{barcodeid}.mapping.log",
            bam = "Result/BWA/Split/{barcodeid}.sortedByPos.rmdp.unique.bam",
            bed = "Result/BWA/Split/{barcodeid}.sortedByPos.rmdp.unique.bed",
        shell:
            "samtools flagstat --threads {threads} {input.sam} > {output.log};"
            "grep -v -e 'XA:Z:' -e 'SA:Z:' {input.sam} | samtools view -F 0x900 -f 0x2 -q 5 -b -o {output.bam};"
            "samtools view {output.bam} -c >> {output.log};"
            "bamToBed -i {output.bam} > {output.bed};"
            "grep 'chrM' {output.bed} -c >> {output.log} || true;"
            "cut -f 1-3 {output.bed} | grep -v 'chrM' | intersectBed -wa -a - -b {input.promoter} -u | wc -l >> {output.log} || true;"
            "cut -f 1-3 {output.bed} | grep -v 'chrM' | intersectBed -wa -a - -b {input.peak} -u | wc -l >> {output.log} || true ;"

    # rule scatac_bulk_unique:
    #     input:
    #         bam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.bam"
    #     output:
    #         unique = "Result/QC/" + config["outprefix"] + "_uniquereads.txt"
    #     threads:
    #         config["cores"]
    #     shell:
    #         "samtools view -@ {threads} -h {input.bam} | grep -v -e 'XA:Z:' -e 'SA:Z:' | samtools view -@ {threads} -F 0x900 -f 0x2 -q 5 -c > {output.unique}"

    # rule scatac_bulk_qcstat:
    #     input:
    #         bam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.bam",
    #         promoter = SCRIPT_PATH + "/annotations/" + config["species"] + "_promoter.bed",
    #         chrM = SCRIPT_PATH + "/annotations/" + config["species"] + "_chrM.bed",
    #         peak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak"
    #     output:
    #         bulkqc = "Result/QC/" + config["outprefix"] + "_bam_stat.txt",
    #         bed = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.ced",
    #     shell:
    #         "samtools flagstat {input.bam} > {output.bulkqc};"
    #         "bamToBed -i {input.bam} > {output.bed};"
    #         "intersectBed -wa -a {output.bed} -b {input.chrM} -u | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | intersectBed -wa -a - -b {input.promoter} -u | wc -l >> {output.bulkqc};"
    #         "cut -f 1-3 {output.bed} | sort -k1,1 -k2,2n | uniq | grep -v 'chrM' | intersectBed -wa -a - -b {input.peak} -u | wc -l >> {output.bulkqc};"

    rule scatac_qcplot:
        input:
            log = qcplot_input,
            fragbed = "Result/QC/" + config["outprefix"] + "_frag.bed",
            # unique = "Result/QC/" + config["outprefix"] + "_uniquereads.txt",
        output:
            stat = "Result/QC/singlecell.txt",
            readdistr = "Result/QC/" + config["outprefix"] + "_scATAC_read_distr.png",
            qcfrag = "Result/QC/" + config["outprefix"] + "_scATAC_fragment_size.png",
            qcmap = "Result/QC/" + config["outprefix"] + "_scATAC_mapping_summary.png",
            qcfrip = "Result/QC/" + config["outprefix"] + "_scATAC_cell_filtering.png",
            validbarcode = "Result/QC/" + config["outprefix"] + "_scATAC_validcells.txt",
        params:
            log = "Result/Log/bamLog/",
            platform = config["platform"],
            outdir = "Result/QC",
            outpre = config["outprefix"],
            stat = "singlecell.txt",
            fragbed = config["outprefix"] + "_frag.bed",
            # unique = config["outprefix"] + "_uniquereads.txt",
        shell:
            "python " + SCRIPT_PATH + "/scATAC_microfluidic_qc.py {params.log} {output.stat};"
            "Rscript " + RSCRIPT_PATH + "/scATACseq_qc.R {params.stat} {params.fragbed} {params.platform} {params.outpre} {params.outdir}"

    rule scatac_allpeakcall:
        input:
            bam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.bam"
        output:
            peak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
            bdg = "Result/Analysis/" + config["outprefix"] + "_all_treat_pileup.bdg"
        params:
            name = config["outprefix"] + "_all"
        log:
            "Result/Log/" + config["outprefix"] + "_macs2_allpeak.log"
        shell:
            "macs2 callpeak -g hs --outdir Result/Analysis -n {params.name} -B -q 0.05 --nomodel --extsize=50 --SPMR -t {input.bam}"

    if config["shortpeaks"]:
        rule scatac_shortfragment:
            input:
                bam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.bam"
            output:
                shortbam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.150bp.bam"
            threads:
                config["cores"]
            shell:
                "samtools view -@ {threads} -h {input.bam} | "
                "awk -F'\\t' 'function abs(x){{return ((x < 0.0) ? -x : x)}} {{if (abs($9)<=150) print}}' | "
                "samtools view -@ {threads} -b -o {output.shortbam}"
        
        rule scatac_shortpeakcall:
            input:
                shortbam = "Result/BWA/" + config["outprefix"] + ".sortedByPos.rmdp.150bp.bam"
            output:
                bed = "Result/Analysis/" + config["outprefix"] + "_150bp_peaks.narrowPeak"
            params:
                name = config["outprefix"] + "_150bp"
            log:
                "Result/Log/" + config["outprefix"] + "_macs2_shortpeak.log"
            shell:
                "macs2 callpeak -g hs --outdir Result/Analysis -n {params.name} -B -q 0.05 --nomodel --extsize=50 --SPMR -t {input.shortbam}"
    
    if config["custompeaks"] and config["shortpeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                shortpeak = "Result/Analysis/" + config["outprefix"] + "_150bp_peaks.narrowPeak",
                custompeak = config["custompeaksloc"]
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} {input.shortpeak} {input.custompeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    elif config["custompeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                custompeaks = config["custompeaksloc"]
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"                
            shell:
                "cat {input.allpeak} {input.custompeaks} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    elif config["shortpeaks"]:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
                shortpeak = "Result/Analysis/" + config["outprefix"] + "_150bp_peaks.narrowPeak"
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} {input.shortpeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"
    else:
        rule scatac_mergepeak:
            input:
                allpeak = "Result/Analysis/" + config["outprefix"] + "_all_peaks.narrowPeak",
            output:
                finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed"
            params:
                catpeaksort = "Result/Analysis/" + config["outprefix"] + "_cat_peaks.bed"
            shell:
                "cat {input.allpeak} "
                "| sort -k1,1 -k2,2n | cut -f 1-4 > {params.catpeaksort};"
                "mergeBed -i {params.catpeaksort} | grep -v '_' | grep -v 'chrEBV' > {output.finalpeak};"
                "rm {params.catpeaksort}"

    rule scatac_countpeak:
        input:
            finalpeak = "Result/Analysis/" + config["outprefix"] + "_final_peaks.bed",
            validbarcode = "Result/QC/" + config["outprefix"] + "_scATAC_validcells.txt"
        output:
            count = "Result/Analysis/" + config["outprefix"] + "_peak_count.h5"
        threads:
            config["cores"]
        params:
            bamdir = "Result/BWA/Split",
            species = config["species"]
        shell:
            "python " + SCRIPT_PATH + "/scATAC_microfluidic_count.py {input.finalpeak} {input.validbarcode} {params.bamdir} {output.count} {threads} {params.species}"

rule scatac_genescore:
    input:
        count = "Result/Analysis/" + config["outprefix"] + "_peak_count.h5"
    output:
        genescore = "Result/Analysis/" + config["outprefix"] + "_genescore.h5"
    params:
        genedistance = config["genedistance"],
        genebed = SCRIPT_PATH + "/annotations/" + config["species"] +"_ensembl.bed",
    shell:
        "python " + SCRIPT_PATH + "/scATAC_genescore.py {input.count} {output.genescore} {params.genedistance} {params.genebed}"

rule scatac_analysis:
    input:
        count = "Result/Analysis/" + config["outprefix"] + "_peak_count.h5",
        genescore = "Result/Analysis/" + config["outprefix"] + "_genescore.h5"
    output:
        specificpeak = "Result/Analysis/" + config["outprefix"] + "_DiffPeaks.tsv",
        clusterplot = "Result/Analysis/" + config["outprefix"] + "_cluster.png",
        annotateplot = "Result/Analysis/" + config["outprefix"] + "_annotated.png",
    params:
        outdir = "Result/Analysis",
        genescore = config["outprefix"] + "_genescore.h5",
        outpre = config["outprefix"],
        count = config["outprefix"] + "_peak_count.h5",
        giggleannotation = config["giggleannotation"],
        species = config["species"],
    threads:
        config["cores"]
    shell:
        "Rscript " + RSCRIPT_PATH + "/scATACseq_pipe.R {params.count} {params.genescore} {params.species} {params.giggleannotation} {params.outpre} {params.outdir} {threads}"

rule scatac_report:
    input:
        # bulkqc = "Result/QC/" + config["outprefix"] + "_bam_stat.txt",
        readdistr = "Result/QC/" + config["outprefix"] + "_scATAC_read_distr.png",
        qcmap = "Result/QC/" + config["outprefix"] + "_scATAC_mapping_summary.png",
        qcfrag = "Result/QC/" + config["outprefix"] + "_scATAC_fragment_size.png",
        qcfrip = "Result/QC/" + config["outprefix"] + "_scATAC_cell_filtering.png",
        count = "Result/Analysis/" + config["outprefix"] + "_peak_count.h5",
        clusterplot = "Result/Analysis/" + config["outprefix"] + "_cluster.png",
        annotateplot = "Result/Analysis/" + config["outprefix"] + "_annotated.png",
        genescore = "Result/Analysis/" + config["outprefix"] + "_genescore.h5"
    output:
        summaryreport = "Result/" + config["outprefix"] + "_scATAC_report.html",
    params:
        outpre = config["outprefix"],
        fastqdir = config["fastqdir"],
        species = config["species"],
        platform = config["platform"]
    shell:
        # "cp {input.readdistr} {input.qcmap} {input.qcfrag} {input.qcfrip} {input.clusterplot} {input.annotateplot} {output.outdir};"
        "python " + SCRIPT_PATH + "/scATAC_report.py {params.outpre} {params.fastqdir} {params.species} {params.platform}"
